### ISSUE TITLE
RFC: Improving SEO with React Helmet 🪖

### ISSUE THREAD

**Issue by simoncrypta** [redwoodjs/redwood#2001]
# The goal is to find how to provide a better SEO out of the box

_This continues the discussions from #1847 for the internalization and also #217 for the conversation about SEO._

## Let's start with meta tags that should be there by default in a Redwood app
_I take inspiration from ["The Ultimate Guide to SEO Meta Tags" ](https://moz.com/blog/the-ultimate-guide-to-seo-meta-tags) blog post_

 To have with every circumstance :
 
- [x] Meta content type
- [x] Viewport
- [x] Title (With htmlWebpackPlugin)
- [x] Meta description

To have with specific circumstances :

- [x] Robots 
- [x] Language
- [ ] Social meta tags
- [ ] Keywords

You can see we have most of the meta tags we need for SEO, but none is dynamic. This means that most of these tags are can be irrelevant. 

### Proposition 1 :

Simply add React Helmet to generate page template with these tags 
```
    <Helmet>
      <title>${pascalName}Page</title>
    </Helmet>
```
keep the other default tags in the HTML file and write documentation for all the specific circumstances

### Proposition 2 :

Create an internal component with all the meta tags like [this Gatsby template ](https://github.com/fabe/gatsby-universal/blob/master/src/components/head/head.js) who are imported into every page with props for dynamic tags and also have a config file for static tags. 

## We need your opinion and proposition

We want a better SEO out of the box for everyone, so if you think we should cover less or more meta tags by default or another way to make it dynamic by pages, I want to know! 

cc @thedavidprice @clairefro

**Comment by simoncrypta** [comment]
To cover the discussion with @clairefro 
adding `<html lang={i18n.language} />` inside Helmet 
would set the present language for `html lang`.

**Comment by dthyresson** [comment]
> Simply add React Helmet to generate page template

I have been using React Helmet, but I have done all the "setup" in a layout and then my page -- or sometimes a control/cell -- if the title is data-driven -- overrides the layout's title (or other meta tags).

This is especially common when showing a bloog page with an article title/author/date and then a social image that is the article's photo, etc.

**Comment by simoncrypta** [comment]
I also currently done the helmet setup in layout with props from pages. I didn't think about data-driven titles, is a good point @dthyresson

**Comment by thedavidprice** [comment]
Huge thanks for getting started on research and this proposal @simoncrypta! This is exactly what we needed. Also, I'm looping in @dac09; he's leading Prerender and has a lot of Jamstack + SEO experience.

Question for everyone: **Is there any reason why we _shouldn't_ have built-in SEO support from the start?** 
- If it was included in generators/boilerplate, hard for me to see a downside (I recognize Redwood, as a dynamic SPA, will likely have a majority of routes behind login)
- I'm assuming I understand the overhead correctly to be installing a package, adding component to Page/Layout, and config

@dthyresson to your comment about using React Helmet within a layout, do you think that will need to change given the new `<Set />` for Routes.js?

**Comment by dac09** [comment]
Helmet is pretty good for this sort of thing, do you guys think we should include it inside redwood? 

I've put up a draft PR here about using helmet for prerendered pages. https://github.com/redwoodjs/redwood/pull/2066/files

Also note, that I'm using react-helmet-async not helmet, because the API is slightly better for SSR

**Comment by thedavidprice** [comment]
- included inside redwood = ✅
- change the title by default for every route = ✅
  - I'm trying to think of a case where I _wouldn't_ want this. So I suggest opt-out

To discuss:
- including in generator(s)
- specific considerations for creating the "right" foundation to then add elements for i18n and a11y (looping in @jtoar)

**Comment by simoncrypta** [comment]
All right, so helmet is now included inside redwood #2909 ✅ thanks @dac09 . Do we wait for another PR to change the title by default for every route by including in page? I think something like this could be valuable to be added by default :

```
    <>
      <Helmet>
        <title>testPage </title>
        <meta name="description" content=" My default route is named test" />
      </Helmet>
      <h1>testPage</h1>
      <p>
        Find me in <code>./web/src/pages/testPage/testPage.tsx</code>
      </p>
      <p>
        My default route is named <code>test</code>, link to me with `
        <Link to={routes.test()}>Test</Link>`
      </p>
    </>
```

> specific considerations for creating the "right" foundation to then add elements for i18n [...]

Also, for i18n, I'm currently testing some configurations to have the right foundation. I think it can wait to see where #2429  will be going for optional path parameters. For now, the best method I find is a React component with a useEffect who listen the i18n useTranslation hook and the path parameter for the language to change the translation and the HTML lang attribute with helmet.

So, @thedavidprice do you think we should close this issue or not?

**Comment by thedavidprice** [comment]
Not done yet!

There are definitely some great, possible next steps here:
1. add defaults to Page generator
2. change the title by default for every route (configurable to opt out)
3. any/all default i18n support

@simoncrypta what do you think about helping out with any/all of these?

@dac09 I know you've talked about making SEO/OG dynamic. Is that something you're considering for v1?

**Comment by simoncrypta** [comment]
I will help with pleasure @thedavidprice, I will open a draft PR tomorrow for all of these. 

When you said 
> any/all default i18n support

Do mean that when i18n is set, the page generator will also add the useTranslation hook and the helmet config for language ?

**Comment by dac09** [comment]
I would love to carve 30 mins out to discuss this with you @simoncrypta - there's a lot of subtleties here that I don't think we can capture over issue chat :) 

- Agreed about page generator default - very happy for you to take this on Simon!
- Changing title by default: this one I'm not sure about. Unlike Gatsby, Redwood doesn't maintain a manifest of pages and their titles and descriptions - so while we can add it to the page generator, how we can determine the titles and descriptions automatically isn't obvious. I'm also not for introducing a page manifest - as I think it caters to a different type of app
- RE: i18n, I think we can do something clever, like you're suggesting. We can detect whether the user has i18n configured by parsing their App.js/tsx and looking for the i18n provider right? Redwood now has built in utilities in `@redwoodj/internal` to help do this easily.

**Comment by dac09** [comment]
@simoncrypta and I had a quick tete-a-tete, and we think we have a way forward with handling titles for page generators, and also making it easier to set the default title of the app through redwood.toml.

This gist:
- add "title" to redwood.toml and automatically use this title instead of having to mess around with webpack
- use this value in the title if possible to for page titles so it has the format {AppName} - {Page Title}

**Comment by dac09** [comment]
Closing this as implemented in #3026 and #2909
